### ISSUE TITLE
Bandwidth tracking

### ISSUE THREAD

**Issue by christian-2** [cilium/cilium#16188]
## Proposal / RFE

This proposal is in relation to new use case and its initial discussion with @rolinh on Slack. It is about determining whether the desired solution is feasible and a possible implementation.

**Is your feature request related to a problem?**

Use case: 
> I have several instances of an application running in a cluster. Each instance consists of a couple of known clients and servers. I’d like to monitor/visualize how much traffic flows between them. One way to approach this would be to aggregate the “sizes” of flows between relevant pairs of pods regardless of client port numbers.

**Describe the solution you'd like**

Flows as reported e.g. by Hubble Relay in its Observer API currently identify who is talking to whom (source/destination) and how (process/errors). They do not yet comprise the "size" of traffic (e.g. byte counts) per flow. The solution would add such information.

**Comment by christian-2** [comment]
I’ve done some coding experiments to see where the extraction and accumulation of byte counts could occur in terms of BPF attachment points. So I have noticed that `struct ct_entry` (as defined in `bpf/lib/common.h` and stored e.g. in `CT_MAP_TCP4`) already has fields `rx_bytes`, `tx_bytes`. What is their exact semantics in relation to this use case and could it perhaps suffice to accumulate those?

**Comment by rolinh** [comment]
Cilium's datapath is aware of each packet size but a Hubble flow does not translate to a network packet one to one because of monitor aggregation. Accumulating individual packet sizes into a flow is actually the difficult part. Changes in @nathanjsweet 's PR #16309 would be sufficient to compute bandwidth if monitor aggregation is set to none. However, for performance reasons, this is not recommended and in a typical deployment several packets are aggregated into a Hubble flow. However, I'm not a datapath expert so I'll defer to @joestringer or @qmonnet to detail an approach to update the datapath to allow implementing full bandwidth tracking.

**Comment by pchaigno** [comment]
I think this is possible since v1.10 with [custom calls](https://github.com/cilium/cilium/blob/master/bpf/custom/README.rst). There's probably not even a lot of changes needed in the example program to achieve this.

**Comment by christian-2** [comment]
Can custom programs call Cilium functions such as  `ipv4_extract_tuple()` and `get_ct_map4()`? If so, a conceivable custom program could obtain a key and map with those and then extract `rx_bytes`, `tx_bytes` (if those are indeed proper starting points for accumulation/aggregation). Or must these programs be simple in the sense that they cannot have dependencies on Cilium’s full code generation engine backend? (I suspect they must/should be simple.)

**Comment by pchaigno** [comment]
> Can custom programs call Cilium functions such as `ipv4_extract_tuple()` and `get_ct_map4()`?

Probably not `get_ct_map4`, but you might be able to call `ipv4_extract_tuple()` by including `conntrack.h`.

**Comment by christian-2** [comment]
I have tried this, but including `bpf/lib/conntrack.h` introduces e.g. a further dependency on `ENDPOINTS_MAP_SIZE`. The only place where I can find its definition (outside tests) is `bpf/node_config.h` and there it says: "This is just a dummy header with dummy values to allow for test compilation without the full code generation engine backend." IIUC the "real" header file is generated by Cilium at runtime whereas I’m currently testing in a simple, "static" development environment. It's basically `git clone` for CIlium plus my custom program added as sibling to `bpf/custom/bytecount.h` and also compiled with `make -C bpf/custom` (as suggested by `bpf/custom/README.rst`). So it seems to me that I cannot make use of functions such as `ipv4_extract_tuple()` from that environment.

It also seems to me that there may be another limitation from a custom program-based approach. IIUC these programs need manual activation by registering them in maps whose pins bear specific endpoint ids (e.g. `/sys/fs/bpf/tc/globals/cilium_calls_custom_1234`). So what’s implemented inside such custom program is not (automatically) effective in the dataplane when new endpoints are created.

**Comment by rolinh** [comment]
> It also seems to me that there may be another limitation from a custom program-based approach.

I think that the custom calls approach proposed by @pchaigno would not be easy to integrate with the Hubble parts.

**Comment by othmane399** [comment]
Hello guys, any news on this topic ?

**Comment by rolinh** [comment]
@othmane399 I'm afraid no one worked on this item (volunteers are welcome!).

**Comment by othmane399** [comment]
Hello @rolinh, I can dig on the subject can I have a recap or a reason for why you're saying that it would not be easy using custom calls ?

**Comment by rolinh** [comment]
> can I have a recap or a reason for why you're saying that it would not be easy using custom calls ?

Right. I'll try to summarize how Hubble gets data from Cilium. You can check Hubble 3/4 parser [here](https://github.com/cilium/cilium/blob/50c72bd5ffec90047e7012ecda8f21ec23efb629/pkg/hubble/parser/threefour/parser.go#L106) for an example of how Hubble reads [the C representation](https://github.com/cilium/cilium/blob/50c72bd5ffec90047e7012ecda8f21ec23efb629/bpf/lib/drop.h#L24) of the structure into [a Go representation](https://github.com/cilium/cilium/blob/50c72bd5ffec90047e7012ecda8f21ec23efb629/pkg/monitor/datapath_drop.go#L22) of it for a drop flow. At a high level, any time there is a [`send_drop_notify`](https://github.com/cilium/cilium/blob/4f27865e975cc095891b68df5f4a7bc0e26ecca5/bpf/lib/drop.h#L33) in the datapath, a struct `drop_notify` is pushed to the perf ring buffer as binary data which Hubble then parses to the Go struct. I don't see an easy way to integrate full bandwidth tracking with a custom call approach in this context.
